compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -600,191 +619,306 @@ impl<'p, 'tcx> FromIterator<PatStack<'p, 'tcx>> for Matrix<'p, 'tcx> {
+                    } else {
+                        // Indices after `arity` are simply shifted
+                        new_subpats.insert(i - arity + 1, sub_set);
+                    }
+                }
+                if !new_subpats_first_col.is_empty() {
+                    new_subpats.insert(0, Seq { subpats: new_subpats_first_col });
+                }
+                Seq { subpats: new_subpats }
             }
+            Alt { .. } => bug!(), // `self` is a patstack
         }
     }
+
+    /// When `self` refers to a patstack that was obtained from splitting an or-pattern, after
+    /// running `unspecialize` it will refer to the original patstack before splitting.
+    ///
+    /// This case is subtle. Consider:
+    /// ```
+    /// match Some(true) {
+    ///     Some(true) => {}
+    ///     None | Some(true | false) => {}
+    /// }
+    /// ```
+    /// Imagine we naively preserved the sets of unreachable subpatterns. Here `None` would return
+    /// the empty set and `Some(true | false)` would return the set containing `true`. Intersecting
+    /// those two would return the empty set, so we'd miss that the last `true` is unreachable.
+    /// To fix that, when specializing a given alternative of an or-pattern, we consider all other
+    /// alternatives as unreachable. That way, intersecting the results will not unduly discard
+    /// unreachable subpatterns coming from the other alternatives. This is what this function does
+    /// (remember that missing entries in the `Alt` case count as full; in other words alternatives
+    /// other than `alt_id` count as unreachable).
+    fn unsplit_or_pat(mut self, alt_id: usize, alt_count: usize, pat: &'p Pat<'tcx>) -> Self {
+        use SubPatSet::*;
+        if self.is_full() {
+            return Full;
+        }
+
+        let set_first_col = match &mut self {
+            Empty => Empty,
+            Seq { subpats } => subpats.remove(&0).unwrap_or(Empty),
+            Full => unreachable!(),
+            Alt { .. } => bug!(), // `self` is a patstack
+        };
+        let mut subpats_first_col = FxHashMap::default();
+        subpats_first_col.insert(alt_id, set_first_col);
+        let set_first_col = Alt { subpats: subpats_first_col, pat, alt_count };
+
+        let mut subpats = match self {
+            Empty => FxHashMap::default(),
+            Seq { subpats } => subpats,
+            Full => unreachable!(),
+            Alt { .. } => bug!(), // `self` is a patstack
+        };
+        subpats.insert(0, set_first_col);
+        Seq { subpats }
+    }
 }
 
 #[derive(Clone, Debug)]
-crate enum Usefulness<'tcx> {
-    /// Pontentially carries a set of sub-branches that have been found to be unreachable. Used
-    /// only in the presence of or-patterns, otherwise it stays empty.
-    Useful(SpanSet),
-    /// Carries a list of witnesses of non-exhaustiveness.
-    UsefulWithWitness(Vec<Witness<'tcx>>),
-    NotUseful,
+enum Usefulness<'p, 'tcx> {
```

Comment:
I've just noticed we don't have a comment for `Usefulness`. I know it's described in the algorithm description at the top of the file, but could you write a brief comment describing what `Usefulness` means and the distinction between the two variants? It doesn't need to be detailed; just a quick overview for a reader skimming the code.